### PR TITLE
Delete cluster/machine implemented.

### PR DESCRIPTION
**Pull request by notjames**
Some relevant code clean-up. Fixed up .gitignore to not include vendor directory or Gopkg.lock, but still needs work.

successfully tested in the following ways:
  * delete just cluster - removes all machines, and cluster secrets
  * delete just machines - removes machines, leaves cluster and secrets intact
  * delete machines first (manually) and then delete cluster (manually) - works as expected

### CODE REVIEW

**Comment by foolusion** [comment]
You need to run 'go fmt'
Also don't gitignore the lock or vendor

File: pkg/controller/machine/machine_controller.go [hunk 1]
```diff
@@ -136,76 +139,61 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
...
 	}
 
 	return reconcile.Result{}, nil
 }
 
 func deleteMachine(r *ReconcileMachine, machine *clusterv1alpha1.CnctMachine) error {
-	if machine.ObjectMeta.Annotations["maas-system-id"] == "" {
-		goto removeFinalizers
+
+	if machine.ObjectMeta.Annotations["maas-system-id"] != "" {
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: This is not the same as the logic before.  Finalizes always need to be removed. Machines are only deleted if the system is is not empty

File: pkg/controller/machine/machine_controller.go [hunk 2]
```diff
@@ -321,7 +305,7 @@ func (r *ReconcileMachine) handleCreate(machine *clusterv1alpha1.CnctMachine, cl
 
 	if isMaster {
 		klog.Info("create kubeconfig")
-		kubeconfig, err := bundle.Kubeconfig(cluster.Name, "https://"+ maasMachine.IPAddresses()[0]+":6443")
+		kubeconfig, err := bundle.Kubeconfig(cluster.Name, "https://"+maasMachine.IPAddresses()[0]+":6443")
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: Use go fmt

File: pkg/cert/gen_cert.go [hunk 3]
```diff
@@ -151,26 +151,26 @@ const (
...
-		c.EtcdKey = value
-	case mapKeyFrontProxy:
-		c.FrontProxy = value
-	case mapKeyFrontProxyKey:
-		c.FrontProxyKey = value
-	case mapKeyK8sClient:
-		c.K8sClient = value
-	case mapKeyK8sClientKey:
-		c.K8sClientKey = value
+    case mapKeyRoot:
```
Comment (foolusion) [inline pkg/cert/gen_cert.go]: Use go fmt

File: pkg/controller/cluster/cluster_controller.go [hunk 4]
```diff
@@ -48,8 +47,8 @@ func Add(mgr manager.Manager) error {
 // newReconciler returns a new reconcile.Reconciler
 func newReconciler(mgr manager.Manager) reconcile.Reconciler {
 	return &ReconcileCluster{
-		Client:        mgr.GetClient(),
-		scheme:        mgr.GetScheme(),
+		Client:  mgr.GetClient(),
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: Use go fmt

File: pkg/controller/cluster/cluster_controller.go [hunk 5]
```diff
@@ -262,28 +237,43 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+		glog.Error(err)
 		return err
 	}
+
 	dataMap := map[string][]byte{}
 	bundle.MergeWithMap(dataMap)
-	secret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "cluster-private-key",
+	secret := &corev1.Secret {
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: Use go fmt

File: pkg/controller/machine/machine_controller.go [hunk 6]
```diff
@@ -270,12 +253,13 @@ func (r *ReconcileMachine) handleCreate(machine *clusterv1alpha1.CnctMachine, cl
 	klog.Infoln("get cluster-private-key secret")
 	secret := corev1.Secret{}
 	if err := r.Client.Get(context.Background(), client.ObjectKey{Name: "cluster-private-key", Namespace: machine.Namespace}, &secret); err != nil {
-		klog.Error("could not get cluster secret data")
-		return reconcile.Result{Requeue: true, RequeueAfter: time.Second}, err
+		klog.Infoln("could not get cluster secret data")
+		//return reconcile.Result{Requeue: true, RequeueAfter: time.Second}, err
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: Why is this return commented out?

File: pkg/controller/machine/machine_controller.go [hunk 7]
```diff
@@ -547,76 +531,10 @@ func doUpgrade(r *ReconcileMachine, machineInstance *clusterv1alpha1.CnctMachine
...
-	if err := r.updateStatus(machineInstance, corev1.EventTypeNormal,
-		common.ResourceStateChange, common.MessageResourceStateChange,
-		machineInstance.GetName(), common.DeletingMachinePhase); err != nil {
-		klog.Errorf("failed to update status for machine %s: %q", machineInstance.GetName(), err)
-	}
-
-	return err
-}
-
 func (r *ReconcileMachine) backgroundRunner(op backgroundMachineOp,
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: This entire function should be removed

File: pkg/controller/machine/machine_controller.go [hunk 8]
```diff
@@ -136,76 +139,61 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 		return r.handleDelete(&machine)
 	}
 
+	//fmt.Printf(" >>>> machine status phase is: '%s'\n", machine.Status.Phase)
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: delete this

File: pkg/controller/machine/machine_controller.go (hunk 1 above)
Comment (notjames) [reply pkg/controller/machine/machine_controller.go]: ok. will fix.

File: pkg/controller/machine/machine_controller.go (hunk 6 above)
Comment (notjames) [reply pkg/controller/machine/machine_controller.go]: Before I commented it out, the secret data never got created the machines never got created. Once I commented this out, things started working again. I commented it, though, as a means of testing, so it can probably just be removed now.

File: pkg/controller/machine/machine_controller.go (hunk 7 above)
Comment (notjames) [reply pkg/controller/machine/machine_controller.go]: Removed it.

File: pkg/controller/machine/machine_controller.go (hunk 8 above)
Comment (notjames) [reply pkg/controller/machine/machine_controller.go]: done

File: pkg/controller/cluster/cluster_controller.go [hunk 9]
```diff
@@ -108,27 +107,42 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 		return reconcile.Result{}, err
 	}
 
+	//fmt.Printf("CLUSTER STATUS IS CURRENTLY: '%s'\n", cluster.Status.Phase)
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: delete this

File: pkg/controller/machine/machine_controller.go (hunk 6 above)
Comment (foolusion) [reply pkg/controller/machine/machine_controller.go]: this section worked before so you must have broken something

Comment (notjames) [reply pkg/controller/machine/machine_controller.go]: I'll check it again

File: pkg/controller/cluster/cluster_controller.go [hunk 10]
```diff
@@ -196,63 +211,21 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
+			// can be removed
+			cluster.Status.Phase = common.DeletedClusterPhase
 			return reconcile.Result{}, r.Update(context.Background(), cluster)
 		}
 	}
 
 	machineList, err := util.GetClusterMachineList(r.Client, cluster.GetName())
 	if err != nil {
-		glog.Errorf("could not list Machines for cluster %s: %q", cluster.GetName(), err)
+		glog.Errorf("could not list Machines for cluster '%s': %q", cluster.GetName(), err)
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: use %q if you want quoted strings

File: pkg/controller/cluster/cluster_controller.go [hunk 11]
```diff
@@ -137,6 +149,7 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 				strings.Join(util.KubernetesVersions(), " "),
 			)
 	}
+	*/
 
 	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
 		// The object is not being deleted, add the finalizer and update the object.
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: This does not need to be done. Finalizer is added on create above

File: pkg/controller/cluster/cluster_controller.go [hunk 12]
```diff
@@ -137,6 +149,7 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 				strings.Join(util.KubernetesVersions(), " "),
 			)
 	}
+	*/
 
 	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: This whole block needs to be above the switch

File: pkg/controller/cluster/cluster_controller.go [hunk 13]
```diff
@@ -196,81 +211,40 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 			}
 
 			// if no Machines left to be deleted
-			// remove our finalizer from the list and update it.
-			cluster.ObjectMeta.Finalizers =
-				util.RemoveString(cluster.ObjectMeta.Finalizers, clusterv1alpha1.ClusterFinalizer)
+			// set phase to deleted so secrets can be deleted and finalizer
+			// can be removed
+			cluster.Status.Phase = common.DeletedClusterPhase
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: we don't need a deleted cluster phase we already have stopping

File: pkg/apis/cluster/common/consts.go [hunk 14]
```diff
@@ -51,6 +51,9 @@ const (
 
 	// The ERROR state indicates the cluster may be unusable
 	ErrorClusterPhase ClusterStatusPhase = "ErrorCluster"
+
+	// The DELETED state indicates the cluster is deleted and the cluster secret(s) should be deleted
+	DeletedClusterPhase ClusterStatusPhase = "DeletedCluster"
```
Comment (foolusion) [inline pkg/apis/cluster/common/consts.go]: This field is redundant. Use Stopping

File: pkg/controller/cluster/cluster_controller.go [hunk 15]
```diff
@@ -124,11 +125,22 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 			common.RunningClusterPhase,
 		)
 		if err != nil {
-			glog.Errorf("could not update cluster %s status: %q", cluster.GetName(), err)
+			glog.Errorf("could not update cluster %q status: %q", cluster.GetName(), err)
 			return reconcile.Result{Requeue: true}, err
 		}
+	case "DeletedCluster":
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: use the const not a string.
also use stopping not deleted

Comment (notjames) [reply pkg/controller/cluster/cluster_controller.go]: Will fix the use of the string vs const. Talking to you OOB about the use of "stopping" --

File: pkg/controller/cluster/cluster_controller.go (hunk 13 above)
Comment (notjames) [reply pkg/controller/cluster/cluster_controller.go]: This needed to be done because I wanted to make sure that the certs are absolutely last in being deleted. This was the way I accomplished that. *shrug*

File: pkg/apis/cluster/common/consts.go [hunk 16]
```diff
@@ -51,6 +51,9 @@ const (
 
 	// The ERROR state indicates the cluster may be unusable
 	ErrorClusterPhase ClusterStatusPhase = "ErrorCluster"
+
+	// The DELETED state indicates the cluster is deleted and the cluster secret(s) should be deleted
+	DeletingClusterPhase ClusterStatusPhase = "DeletingCluster"
```
Comment (foolusion) [inline pkg/apis/cluster/common/consts.go]: delete this

File: pkg/controller/cluster/cluster_controller.go [hunk 17]
```diff
@@ -196,81 +153,93 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 			}
 
 			// if no Machines left to be deleted
-			// remove our finalizer from the list and update it.
-			cluster.ObjectMeta.Finalizers =
-				util.RemoveString(cluster.ObjectMeta.Finalizers, clusterv1alpha1.ClusterFinalizer)
-			return reconcile.Result{}, r.Update(context.Background(), cluster)
+			// set phase to deleted so secrets can be deleted and finalizer
+			// can be removed
+			cluster.Status.Phase = common.DeletingClusterPhase
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: use stopping

File: pkg/controller/cluster/cluster_controller.go [hunk 18]
```diff
@@ -196,81 +153,93 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
...
-			cfg, err := machine.NewCmdConfig(r.Client, masterMachine, privateKeySecret.Data["private-key"])
-			if err != nil {
-				return reconcile.Result{}, err
-			}
-			// get kubeconfig
-			kubeConfig, err := machine.GetKubeConfig(cfg, nil)
-			if err != nil {
-				return reconcile.Result{}, err
-			}
+	/* TODO: determine parts of this we need
```
Comment (foolusion) [inline pkg/controller/cluster/cluster_controller.go]: just delete this

File: pkg/apis/cluster/common/consts.go (hunk 14 above)
Comment (notjames) [reply pkg/apis/cluster/common/consts.go]: Discussed OOB - changed to deleting.

File: pkg/controller/cluster/cluster_controller.go (hunk 11 above)
Comment (notjames) [reply pkg/controller/cluster/cluster_controller.go]: This was fixed I believe.

File: pkg/controller/machine/machine_controller.go [hunk 19]
```diff
@@ -337,8 +322,10 @@ func (r *ReconcileMachine) handleCreate(machine *clusterv1alpha1.CnctMachine, cl
 	// update status to "creating"
 	machine.Status.Phase = common.ReadyMachinePhase
 	machine.Status.KubernetesVersion = cluster.Spec.KubernetesVersion
-	machine.ObjectMeta.Annotations["maas-ip"] = createResponse.IPAddresses[0]
-	machine.ObjectMeta.Annotations["maas-system-id"] = createResponse.SystemID
+
+	machine.ObjectMeta.Annotations["maas-ip"] = maasMachine.IPAddresses()[0]
+	machine.ObjectMeta.Annotations["maas-system-id"] = maasMachine.SystemID()
+	machine.ObjectMeta.Annotations["maas-system-name"] = maasMachine.FQDN()
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: why did you add this?

File: pkg/controller/machine/machine_controller.go [hunk 20]
```diff
@@ -150,62 +149,49 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
...
-	systemID := machine.ObjectMeta.Annotations["maas-system-id"]
-	if systemID == "" {
-		goto removeFinalizers
-	}
-
-	if err := r.MAASClient.Delete(context.Background(), &maas.DeleteRequest{SystemID: systemID}); err != nil {
-		return errs.Wrapf(err, "could not delete machine %s with system id %q", machine.Name, *machine.Spec.ProviderID)
+	if machine.ObjectMeta.Annotations["maas-system-id"] != "" {
+		if err := r.MAASClient.Delete(context.Background(), nil, machine); err != nil {
+			return errs.Wrapf(err, "could not delete machine %q with system id %q (%s)", machine.Name, *machine.Spec.ProviderID, machine.ObjectMeta.Annotations["maas-system-name"])
```
Comment (foolusion) [inline pkg/controller/machine/machine_controller.go]: you don't need system name, you can search for the systemid in maas

Comment (notjames) [reply pkg/controller/machine/machine_controller.go]: too much time is required for an operator to search for a machine by maas id vs looking at the maas gui with the maas FQDN. I believe this is a quick way to get an operator very important information at low cost.
